index.ts

```diff
@@ -0,0 +1,74 @@
+import { EventEmitter } from 'events';
+
+type Handler = (...args: any[]) => void;
+type EventMap = {
+  [k: string]: Handler | Handler[] | undefined;
+};
+
+function safeApply<T, A extends any[]> (handler: (this: T, ...args: A) => void, context: T, args: A): void {
+  try {
+    Reflect.apply(handler, context, args);
+  } catch (err) {
+    // Throw error after timeout so as not to interrupt the stack
+    setTimeout(() => {
+      throw err;
+    });
+  }
+}
+
+function arrayClone<T> (arr: T[]): T[] {
+  const n = arr.length;
+  const copy = new Array(n);
+  for (let i = 0; i < n; i++) {
+    copy[i] = arr[i];
+  }
+  return copy;
+}
+
+export default class SafeEventEmitter extends EventEmitter {
+  emit (type: string, ...args: any[]): boolean {
+    let doError = type === 'error';
+
+    const events: EventMap = (this as any)._events;
+    if (events !== undefined) {
+      doError = doError && events.error === undefined;
+    } else if (!doError) {
+      return false;
+    }
+
+    // If there is no 'error' event listener then throw.
+    if (doError) {
+      let er;
+      if (args.length > 0) {
+        er = args[0];
+      }
+      if (er instanceof Error) {
+        // Note: The comments on the `throw` lines are intentional, they show
+        // up in Node's output if this results in an unhandled exception.
+        throw er; // Unhandled 'error' event
+      }
+      // At least give some kind of context to the user
+      const err = new Error('Unhandled error.' + (er ? ' (' + er.message + ')' : ''));
```

Comment:
Nit: This might be more readable using a string template?

Reply:
Seconded, but non-blocking.